### PR TITLE
MAPSJS-2953: Support custom DataAdapters for VectorTileDecoder.

### PR DESCRIPTION
**Pull request by atomicsulfate**
Signed-off-by: Andres Mandado <andres.mandado-almajano@here.com>

VectorTileDecoder has now a method getDataAdapter() whose default
implementation returns one of the 3 default adapters (omv, geojson,
or geojson-vt), the one that supports the provided data.

The method can be overriden by a subclass to provide a custom adapter:

```javascript
class CustomDataAdapter implements DataAdapter{
    // This method can be used to select an appropiate adapter for the given data when a decoder is meant to support,
    // multiple formats. It may just return true if there's only one supported format.
    // @override
    canProcess(data: ArrayBufferLike | {}): boolean { return true;} 

     // @override
    process(data: ArrayBufferLike | {},  decodeInfo: DecodeInfo,  geometryProcessor: IGeometryProcessor): void {
         // TODO
   }
}

class CustomVectorTileDecoder extends VectorTileDecoder {
    private readonly m_customDataAdapter = new CustomDataAdapter();
/* @override */
getDataAdapter(data: ArrayBufferLike | {}): DataAdapter | undefined {
        return this.m_customDataAdapter;
    }
}
````

### CODE REVIEW

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 1]
```diff
@@ -317,11 +300,16 @@ export class VectorTileDecoder extends ThemedTileDecoder {
         projection: Projection
     ): Promise<DecodedTile> {
         const startTime = PerformanceTimer.now();
+        const dataAdapter = this.getDataAdapter(data);
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:303]: Does this need to be called for every tile key? Could it be lazy initialized? Perhaps not if the class has to support different formats at run time.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 2]
```diff
@@ -56,28 +56,24 @@ export class VectorTileDataProcessor implements IGeometryProcessor {
...
-
-        // Register the default adapters.
-
-        const omvDataAdapter = new OmvDataAdapter(this, dataPreFilter, logger);
-        omvDataAdapter.roundUpCoordinatesIfNeeded = m_roundUpCoordinatesIfNeeded;
-        this.m_dataAdapters.push(omvDataAdapter);
-
-        this.m_dataAdapters.push(new GeoJsonVtDataAdapter(this, dataPreFilter, logger));
-        this.m_dataAdapters.push(new GeoJsonDataAdapter(this, dataPreFilter, logger));
+        if (this.m_dataAdapter instanceof OmvDataAdapter) {
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:69]: Can this be done outside this class?

I find it a bit ugly that VectorTileDataProcessor has to know about the OmvDataAdapter internals.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 3]
```diff
@@ -125,7 +102,7 @@ export class VectorTileDataProcessor implements IGeometryProcessor {
             this.m_languages
         );
 
-        dataAdapter.process(data, decodeInfo);
+        this.m_dataAdapter.process(data, decodeInfo, this);
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:105]: Not related to your code, but this is a bit weird to have a `this` pointer.

It shows that the `VectorTileDataProcessor` does two things. It is a geometry processor and it also does the wireup between DataAdapter and the DecodedTileEmitter.

I think the IGeometryProcessor should be separated into a separate class from the VectorTileDataProcessor, but this can be done later.

File: @here/harp-vectortile-datasource/lib/DataAdapter.ts [hunk 4]
```diff
@@ -5,22 +5,19 @@
...
-     * `DataAdapter`'s id.
-     */
-    id: string;
-
     /**
      * Checks if the given data can be processed by this `DataAdapter`.
      *
      * @param data - The raw data to adapt.
      */
+
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/DataAdapter.ts:20]: done

File: @here/harp-vectortile-datasource/lib/DataAdapter.ts [hunk 5]
```diff
@@ -29,5 +26,9 @@ export interface DataAdapter {
      * @param data - The raw data to process.
      * @param decodeInfo - The `DecodeInfo` of the tile to process.
      */
-    process(data: ArrayBufferLike | {}, decodeInfo: DecodeInfo): void;
+    process(
+        data: ArrayBufferLike | {},
+        decodeInfo: DecodeInfo,
+        geometryProcessor: IGeometryProcessor
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/DataAdapter.ts:32]: done

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 1 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:301]: Changed to do lazy initialization.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 6]
```diff
@@ -427,6 +413,25 @@ export class VectorTileDecoder extends ThemedTileDecoder {
         }
     }
 
+    /**
```
Comment (germanz) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:416]: MInor spelling fixes.

```
    /**
     * Returns the appropriate data adapter to convert the given data into the format expected by
     * VectorTileDecoder.
     * @note Default adapters are available for GeoJson and OMV formats.
     * Child classes may override this function to support additional formats.
     *
     * @param data - The input data to be converted.
     * @returns The DataAdapter to convert the data, or undefined if there's no adapter for that
     * data format.
     */
```

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 7]
```diff
@@ -111,149 +113,145 @@ export class VectorTileDataProcessor implements IGeometryProcessor {
...
-            }
-        }
-
-        const techniques = this.applyKindFilter(
-            this.m_styleSetEvaluator.getMatchingTechniques(env, layer, "line"),
-            GeometryKind.Line
+        assert(this.m_decodedTileEmitter !== undefined);
+        // Pass feature modifier method to processFeature if there's any modifier. Get it from any
+        // modifier, processFeature will later apply it to all using Function.apply().
+        const featureModifierFunc = this.m_featureModifiers?.[0].doProcessLineFeature;
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:147]: Why do you use just the first feature modifier function? It is an array so there could be multiple ones?

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:69]: I've moved all OMV-specific decode options to OmvDataAdapter.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 7 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:147]: It explains in the comment above, to get a method you need to get it from any instance of the class, it doesn't matter which. Later I'm binding that method to different objects using Function.apply(). 
Anyway, I've refactored the code and now that's not there anymore. I've moved the code to OmvDataAdapter.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 3 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:86]: Yes, makes sense. But I'd rather leave it for a separate PR.

File: @here/harp-mapview-decoder/lib/ThemedTileDecoder.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ export abstract class ThemedTileDecoder implements ITileDecoder {
     }
 
     decodeTile(
-        data: ArrayBufferLike,
+        data: ArrayBufferLike | {},
```
Comment (nzjony) [inline @here/harp-mapview-decoder/lib/ThemedTileDecoder.ts:38]: I already have a PR for this: https://github.com/heremaps/harp.gl/pull/2198/files and I invited you to review, but I guess you missed it.

File: @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts [hunk 9]
```diff
@@ -56,11 +56,6 @@ export interface ILineGeometry {
  * line and polygon geometries, see [[VectorTileDecoder]].
  */
 export interface IGeometryProcessor {
-    /**
-     * The offset to apply the storage level of a tile to compute the zoom level.
-     */
-    storageLevelOffset?: number;
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts:63]: should this be first deprecated?

File: @here/harp-vectortile-datasource/lib/OmvDataFilter.ts [hunk 10]
```diff
@@ -713,7 +713,7 @@ export class ComposedDataFilter implements OmvFeatureFilter {
      */
     get hasKindFilter() {
         return this.filters.reduce<boolean>(
-            (result, filter) => result && filter.hasKindFilter,
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/OmvDataFilter.ts:716]: should be a separate commit, it seems unrelated to the rest of the PR

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts [hunk 11]
```diff
@@ -226,6 +226,17 @@ enum LineType {
 type TexCoordsFunction = (tilePos: THREE.Vector2, tileExtents: number) => THREE.Vector2;
 const tmpColor = new THREE.Color();
 
+/**
+ * Options for VectorTileDataEmitter, @see {@link DecoderOptions} and {@link OmvDecoderOptions}.
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:230]: This doesn't seem to be rendered correctly, please try this locally and fix.

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts [hunk 12]
```diff
@@ -226,6 +226,17 @@ enum LineType {
...
+ * @internal
+ */
+export interface VectorTileDataEmitterOptions {
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:233]: Ideally this should be in a separate commit / PR, because it seems just a refactoring and not related to the actual ticket?

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts [hunk 13]
```diff
@@ -242,14 +253,20 @@ export class VectorTileDataEmitter {
     private m_maxGeometryHeight: number = 0;
     private m_minGeometryHeight: number = 0;
 
+    private readonly m_gatherFeatureAttributes: boolean;
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:256]: Why don't you just store the options object here instead of repeating the instance methods? At least that is something that we seem to do regularly in other classes.

File: @here/harp-mapview-decoder/lib/ThemedTileDecoder.ts (hunk 8 above)
Comment (atomicsulfate) [reply @here/harp-mapview-decoder/lib/ThemedTileDecoder.ts:38]: Oh, sorry, I didn't see that. I just merged it.

Comment (nzjony) [reply @here/harp-mapview-decoder/lib/ThemedTileDecoder.ts:38]: No worries.

File: @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts (hunk 9 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts:63]: This wasn't part of the API, it's not exported in the package index.ts. Now it's only exposed indirectly because it's a parameter type of DataAdapter interface.

File: @here/harp-vectortile-datasource/lib/OmvDataFilter.ts (hunk 10 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/OmvDataFilter.ts:716]: I had to add a lot of missing unit tests for VectorTileDecoder, VectorTileDataProcessor, OmvDataAdapter... I found this bug while unit testing VectorTileDataProcessor, that's why I fixed it here.
It could be moved to a separate PR, I'd have to move the failing test as well. But since it's a one liner, is it ok if I leave it here?

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts (hunk 11 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:230]: DecoderOptions works. OmvDecoderOptions doesn't because it's marked as "@hidden". I've changed it to "@internal" instead, so that we can see it in our documentation but still can omit it from public documentation by not exporting internals.

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts (hunk 12 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:233]: True. But since it hardly adds any complexity to the change, may we keep it here? Otherwise I have to rewrite all the usages back to the original signature and then introduce the interface again in a separate PR.

I see you try to reduce the size of this PR, but most of it comes from the unit tests (+600 lines) and then moving OMV specific options out of VectorTileDecoder and VectorTileDataProcessor.
The changes related to custom adapters are in the first two commits. The rest of the commits are unit tests and refactoring.

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts (hunk 13 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:256]: makes sense. Wouldn't you rather do this in a separate PR?

File: @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts (hunk 9 above)
Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/IGeometryProcessor.ts:63]: Good point, I should also check the index.ts...

File: @here/harp-vectortile-datasource/lib/OmvDataFilter.ts (hunk 10 above)
Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/OmvDataFilter.ts:716]: All good, ok, sounds fine.

File: @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts (hunk 12 above)
Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/VectorTileDataEmitter.ts:233]: Ok, fair enough.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 14]
```diff
@@ -311,21 +264,22 @@ export class VectorTileDecoder extends ThemedTileDecoder {
         }
         const dataAdapter = this.m_dataAdapter!;
         assert(dataAdapter.canProcess(data));
+        if (this.m_optionsChanged) {
+            if (dataAdapter instanceof OmvDataAdapter) {
+                const omvOptions = this.m_options as OmvDecoderOptions;
+                dataAdapter.configure(omvOptions, styleSetEvaluator);
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:270]: why not put this method into the DataAdapter interface (Open-closed principle if I'm not mistaken)...?

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts [hunk 15]
```diff
@@ -337,122 +291,32 @@ export class VectorTileDecoder extends ThemedTileDecoder {
...
-                this.m_enableElevationOverlay = omvOptions.enableElevationOverlay;
-            }
-            if (omvOptions.roundUpCoordinatesIfNeeded !== undefined) {
-                this.m_roundUpCoordinatesIfNeeded = omvOptions.roundUpCoordinatesIfNeeded;
-            }
-        }
-        if (options?.languages !== undefined) {
-            this.languages = options.languages;
-        }
+        this.m_options = {
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:294]: Minor, you could merge the `m_options` and `m_optionsChanged` into a class and then add a method update which would encapsulate the `m_optionsChanged = true` change.

File: @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts [hunk 16]
```diff
@@ -248,20 +345,25 @@ export class OmvDataAdapter implements DataAdapter, OmvVisitor {
      */
     visitPointFeature(feature: com.mapbox.pb.Tile.IFeature): void {
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:346]: this should be void | undefined

File: @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts [hunk 17]
```diff
@@ -460,27 +512,58 @@ export class OmvDataAdapter implements DataAdapter, OmvVisitor {
...
+        }
+
+        const properties = readAttributes(this.m_layer, feature);
+        const env = new MapEnv(properties);
+        if (
+            this.m_featureModifiers?.find(fm => {
+                // TODO: The logic of feature ignore should be actually in the feature filtering
+                // mechanism - see OmvFeatureFilter.
+                assert(modifierFunc !== undefined);
+                return !modifierFunc!.apply(fm, [layerName, env, this.m_tileKey.level]);
```
Comment (nzjony) [inline @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:555]: this is some genius level typescript, I tried for an hour to figure it out and I just understand...

Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:555]: Well, I found it is possible to do it by setting:

modifierName: "doProcessPointFeature" | "doProcessLineFeature" | "doProcessPolygonFeature",

And then go `fm[modifierName]`

Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:555]: However.... that isn't so nice to have to hard code the methods there. I tried to find a way to extract it from the OmvFeatureModifier but to no avail

Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:555]: Nevermind,

Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:555]: ok

File: @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts (hunk 16 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/adapters/omv/OmvDataAdapter.ts:346]: I removed the "return undefined".

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 14 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:270]: Yes, it's kind of hacky. If we only had a configure method getting an option map, I'd move it to DataAdapter interface. But there's two more things specific of OMV:
1. You have to pass the StyleSetEvaluator to create a StyleSetDataFilter.
2. You have to retrieve back a OmvFeatureFilter created by OmvDataAdapter to use it to filter kinds in VectorTileDataProcessor (see a couple of lines below the call to configure).

These details are so far specific of OMV. We could generalize them for all formats, but the question is if we really want to do it:
- Kinds filtering is deprecated (see BaseTechniqueParams.kind), eventually we'll remove it.
- All kinds of filtering could be implemented directly in the style, using dynamic properties if necessary.
- FeatureModifiers where introduced for political views, but I think it should be doable via styling  as well.

I think this needs a revisit for further refactoring, maybe when we add OCM.

I can put a comment and create a ticket for this.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 15 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts]: done.

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 14 above)
Comment (nzjony) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:269]: @atomicsulfate , you could theoretically pass the `This` pointer, so the data adapter could extract the options / stylesetevaluator internally and not here.

However I get your point, if this is mostly things that can be deprecated, then a comment + ticket should be enough to let future developers no that this shouldn't be permanent.

**Review by nzjony — APPROVED** [review]

File: @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts (hunk 14 above)
Comment (atomicsulfate) [reply @here/harp-vectortile-datasource/lib/VectorTileDecoder.ts:269]: Ticket created
